File: lib/swimCoinClaims.js
import { coreMetricsMatch } from './swimDuplicates.js';

/** Total coins awarded when this session was saved. */
export const sessionTotalCoins = (session) =>
  (session?.coinsEarned || 0) + (session?.coinBonus || 0);

/** Fingerprint stored when a session is deleted so coins cannot be re-earned. */
export const createCoinClaim = (session) => ({
  date: session.date,
  metrics: {
    distanceM: session.metrics?.distanceM ?? null,
    durationSec: session.metrics?.durationSec ?? null,
    paceSecPer100m: session.metrics?.paceSecPer100m ?? null,
    timeRange: session.metrics?.timeRange || '',
  },
});

export const findSpentCoinClaim = (claims, candidate) => {
  if (!claims?.length || !candidate?.date) return null;
  return claims.find((claim) => coreMetricsMatch(claim, candidate)) ?? null;
};

/** @internal */
export const __testing = { sessionTotalCoins, createCoinClaim };

File: ios/AapSC/Lib/SwimCoinClaims.swift
import Foundation

enum SwimCoinClaims {
    static func sessionTotalCoins(_ session: SwimSession) -> Int {
        (session.coinsEarned ?? 0) + (session.coinBonus ?? 0)
    }

    static func createCoinClaim(_ session: SwimSession) -> SpentCoinClaim {
        SpentCoinClaim(
            date: session.date,
            metrics: ClaimMetrics(
                distanceM: session.metrics.distanceM,
                durationSec: session.metrics.durationSec,
                paceSecPer100m: session.metrics.paceSecPer100m,
                timeRange: session.metrics.timeRange
            )
        )
    }

    static func findSpentCoinClaim(_ claims: [SpentCoinClaim], candidate: SwimSession) -> SpentCoinClaim? {
        guard !claims.isEmpty else { return nil }
        return claims.first { SwimDuplicates.coreMetricsMatch($0, candidate) }
    }
}
